ci/do_ci.sh

```diff
@@ -125,19 +125,24 @@ elif [[ "$1" == "bazel.debug.server_only" ]]; then
   exit 0
 elif [[ "$1" == "bazel.asan" ]]; then
   setup_clang_toolchain
-  echo "bazel ASAN/UBSAN debug build with tests..."
+  echo "bazel ASAN/UBSAN debug build with tests"
+  echo "Building and testing envoy tests..."
+  cd "${ENVOY_SRCDIR}"
+  bazel_with_collection test ${BAZEL_TEST_OPTIONS} -c dbg --config=clang-asan //test/...
```

Comment:
How come we can't build/test `@envoy` directly anymore? I get that we're in a different workspace etc, but don't see why we should be able to use these paths.

Reply:
It's a combination of a few things:
1. `bazel` test [automatically sets](https://docs.bazel.build/versions/master/test-encyclopedia.html#initial-conditions) the TEST_WORKSPACE environment variable to the local repository's workspace name.  
   - Without this PR, when `bazel test` is run from $ENVOY_FILTER_EXAMPLE_SRCDIR, TEST_WORKSPACE would evaluate to "envoy", since that was the workspace name set by WORKSPACE.filter.example.
   - With this PR, when `bazel test` is run from $ENVOY_FILTER_EXAMPLE_SRCDIR, TEST_WORKSPACE evaluates to "envoy_filter_example".

2. TEST_WORKSPACE affects TEST_RUNDIR. Here's where envoy/test/main.cc [sets TEST_RUNDIR](https://github.com/envoyproxy/envoy/blob/b52452e32502e288de8b7e61141b75f99f3fa2f6/test/main.cc#L26) to $TEST_SRCDIR/$TEST_WORKSPACE. And TEST_RUNDIR is how Envoy::TestEnvironment finds test input data: TestEnvironment::runfilesPath() just [appends the path](https://github.com/envoyproxy/envoy/blob/b52452e32502e288de8b7e61141b75f99f3fa2f6/test/test_common/environment.h#L92) it is given to TEST_RUNDIR.

3. Empirically, bazel places runfiles for build targets in $TEST_SRCDIR/<workspace-name>, where <workspace-name> is the name of the (maybe external) workspace containing the build target. With this PR, prior to the latest changes that split `bazel test` invocations, runfiles for \@envoy tests were in $TEST_SRCDIR/envoy, while `bazel test` invoked from $ENVOY_FILTER_EXAMPLE_SRCDIR was setting TEST_RUNDIR to $TEST_SRCDIR/envoy_filter_example and looking for them there.

Splitting the `bazel test` invocations ensures that TEST_WORKSPACE is always the same as the <workspace-name> in which the runfiles for tests can be found. This issue didn't arise before, since we used the same workspace name "envoy" in both cases.

Reply:
This seems to be masking an underlying issue in how we or Bazel are setting up the rundir path for tests. Shouldn't we be able to `bazel test @foo//:bar` for some external repository?

Reply:
Here is a very basic, self-contained repro that illustrates the behavior: https://gist.github.com/mpwarres/b83539d17f92d63d9c25794a84a734ab

@dkelmer / @dslomov, is this supposed to work, or is running a test with runfiles that belongs to external dependency A from some other workspace B not a supported use case?

@htuch I'm not sure how else to address the Bazel behavior within the confines of this PR, any suggestions welcome. In some sense the Bazel behavior was masked even prior to this PR, by virtue of the external dependency having the same name as the workspace using it.

Reply:
@htuch from a repository consuming envoy as deps (e.g. envoy-filter-example, istio/proxy), bazel test @envoy//test/... will fail some of tests depending on TEST_WORKSPACE, I think it is a separate issue which we should address, but should not block on this PR.

Reply:
@lizan yep, I've already merged this PR. I'm hoping we can followup with an issue on bazelbuild GH wit @mpwarres 's repro.